forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabRepository.java

```diff
@@ -313,6 +313,31 @@ public Optional<String> fileContents(String filename, String ref) {
         return Optional.of(new String(decodedContent, StandardCharsets.UTF_8));
     }
 
+    @Override
+    public void writeFileContents(String content, String filename, Branch branch, String message, String authorName, String authorEmail) {
+        var encodedFileName = URLEncoder.encode(filename, StandardCharsets.UTF_8);
+        var body = JSON.object()
+                .put("commit_message", message)
+                .put("branch", branch.name())
+                .put("author_name", authorName)
+                .put("author_email", authorEmail)
+                .put("encoding", "base64")
+                .put("content", new String(Base64.getEncoder().encode(content.getBytes(StandardCharsets.UTF_8)), StandardCharsets.UTF_8));
+        request.put("repository/files/" + encodedFileName)
+                .body(body)
+                .onError(response -> {
+                    // Gitlab requires POST for creating new files and PUT for updating existing.
+                    // Retry with POST if we get 400 response.
+                    if (response.statusCode() == 400) {
+                        return Optional.of(request.post("repository/files/" + encodedFileName)
+                                .body(body)
+                                .execute());
```

Comment:
You don't need an `onError` to return `Optional.empty()` for the POST case? Or is it the default if you don't specify an error handler?

Reply:
The default for `execute()` is to throw `UncheckedRestException` if we get any kind of error response from the server. Here I made the assumption that replacing files is more common than creating new files, and simply first try PUT to replace a file and if that fails with the documented error code 400, we retry with POST to create a new file. If the POST fails, then we just let it fail.

Another way of doing it would have been to first call GET to see if the file exists and then pick either PUT or POST depending on the result. I just tried to keep the number of calls down for the current common case.